fw/cache.c

```diff
@@ -1784,8 +1784,14 @@ tfw_cache_add(TfwHttpResp *resp, tfw_http_cache_cb_t action)
 	 */
 
 out:
-	resp->msg.ss_flags |= keep_skb ? SS_F_KEEP_SKB : 0;
-	action((TfwHttpMsg *)resp);
+	if (likely((req->cache_ctl.flags & TFW_HTTP_CC_CACHE_PURGE) == 0)) {
+		resp->msg.ss_flags |= keep_skb ? SS_F_KEEP_SKB : 0;
+		action((TfwHttpMsg *)resp);
+	}
+	else {
+		tfw_http_conn_msg_free((TfwHttpMsg *)resp);
+		tfw_http_send_resp(req, 200, "purge: success");
```

Comment:
We do our best to to minimize memory allocations and copyings but here you just remove the original response and create a new one from scratch. It's more efficient to just drop body from the response. Probably Content-Lenght and Transfer-Encoding headers as well. See the big comment about the proper protocol

Reply:
The usual Tempesta FW HTTP response for PURGE looks like
```
< HTTP/1.1 200 OK
< date: Sun, 26 Sep 2021 13:20:58 GMT
< content-length: 0
< server: Tempesta FW/0.7.0
```
and the underlying server (Nginx in my current settings) returns
```
< HTTP/1.1 200 OK
< Server: nginx/1.10.3
< Date: Sun, 26 Sep 2021 13:19:54 GMT
< Content-Type: text/html
< Content-Length: 3378
< Last-Modified: Thu, 02 Sep 2021 20:40:13 GMT
< Connection: keep-alive
< ETag: "613136ad-d32"
< Expires: Sun, 26 Sep 2021 13:29:54 GMT
< Cache-Control: max-age=600
< Cache-Control: public
< Accept-Ranges: bytes
<
```
I remove body and rewrite Content-Length to zero, leaving all the rest headers intact.

Reply:
I committed changes for this issue.